### ISSUE TITLE
Add short circuit for public method to avoid unnecessary synchronization and cache

### ISSUE THREAD

**Issue by quaff** [jkuhnert/ognl#90]
Here is my proposal:
add
```
         if (Modifier.isPublic(method.getModifiers()) && Modifier.isPublic(method.getDeclaringClass().getModifiers()))
         {
     			// real public method should invoke directly
     			return method.invoke(target, argsArray);
         }
```
before 
https://github.com/jkuhnert/ognl/blob/32d076d209c11da12bebe48d88321b961ac1b087/src/java/ognl/OgnlRuntime.java#L1132

@yasserzamani @lukaszlenart  WDYT?

**Comment by yasserzamani** [comment]
It at least misses two things: checkPermission and invokeMethodInSandbox.

**Comment by quaff** [comment]
Please think twice, is it necessary for real public methods? if yes, we could call it before real method invoke, my point is avoiding synchronization and reduce cache size.

**Comment by yasserzamani** [comment]
Yes I see; I meant your proposal deletes two functionalities from current logic for public methods: 1. checkPermision: checks if user has blocked this method execution. 2. invokeMethodInSandbox: runs user method inside internal sandbox.

Anyway, do you have any problem with its performance currently?

________________________________________

**Comment by quaff** [comment]
I've tested one simple scene of my actual application, throughput could increase about 5%, I think it worth a shot.

**Comment by JCgH4164838Gh792C124B5** [comment]
Hello.

The proposed change might improve performance (avoiding some synchronization) for public method calls but there may not be a clean way to preserve the permission checks and sandboxing calls in the new code path (without reintroducing some form of synchronization).  Some kind of opt-in mechanism to bypass the checks/synchronization for public methods might be possible ... but the risk of bypassing the checks probably outweighs a small performance gain.

**Comment by KaporMirr** [comment]
Předám dále....
